src/components/tabs/tabs.js

```diff
@@ -0,0 +1,272 @@
+  var $tabs = this.$tabs
+  var $tabList = $module.querySelector('.govuk-tabs__list')
+  var $tabListItems = $module.querySelectorAll('.govuk-tabs__list-item')
+
+  if (!$tabs || !$tabList || !$tabListItems) {
+    return
+  }
+
+  $tabList.removeAttribute('role')
+
+  nodeListForEach($tabListItems, function ($item) {
+    $item.removeAttribute('role', 'presentation')
+  })
+
+  nodeListForEach($tabs, function ($tab) {
+    // Remove events
+    $tab.removeEventListener('click', $tab.boundTabClick, true)
+    $tab.removeEventListener('keydown', $tab.boundTabKeydown, true)
+
+    // Unset HTML attributes
+    this.unsetAttributes($tab)
+  }.bind(this))
+
+  // Remove hashchange event handler
+  window.removeEventListener('hashchange', $module.boundOnHashChange, true)
+}
+
+Tabs.prototype.onHashChange = function (e) {
+  var hash = window.location.hash
+  if (!this.hasTab(hash)) {
+    return
+  }
+  // Prevent changing the hash twice from the same action
+  if (this.changingHash) {
+    this.changingHash = false
+    return
+  }
+
+  // Show either the active tab according to the URL's hash or the first tab
+  var $previousTab = this.getCurrentTab()
+  var $activeTab = this.getTab(hash) || this.$tabs[0]
+
+  this.hideTab($previousTab)
+  this.showTab($activeTab)
+  $activeTab.focus()
+}
+
+Tabs.prototype.hasTab = function (hash) {
+  return this.$module.querySelector(hash)
+}
+
+Tabs.prototype.hideTab = function ($tab) {
+  this.unhighlightTab($tab)
+  this.hidePanel($tab)
+}
+
+Tabs.prototype.showTab = function ($tab) {
+  this.highlightTab($tab)
+  this.showPanel($tab)
+}
+
+Tabs.prototype.getTab = function (hash) {
+  return this.$module.querySelector('a[role="tab"][href="' + hash + '"]')
+}
+
+Tabs.prototype.setAttributes = function ($tab) {
+  // set tab attributes
+  var panelId = this.getHref($tab).slice(1)
+  $tab.setAttribute('id', 'tab_' + panelId)
+  $tab.setAttribute('role', 'tab')
+  $tab.setAttribute('aria-controls', panelId)
+  $tab.setAttribute('tabindex', '-1')
+
+  // set panel attributes
+  var $panel = this.getPanel($tab)
+  $panel.setAttribute('role', 'tabpanel')
+  $panel.setAttribute('aria-labelledby', $tab.id)
+  $panel.classList.add(this.jsHiddenClass)
+}
+
+Tabs.prototype.unsetAttributes = function ($tab) {
+  // unset tab attributes
+  $tab.removeAttribute('id')
+  $tab.removeAttribute('role')
+  $tab.removeAttribute('aria-controls')
+  $tab.removeAttribute('tabindex')
+
+  // unset panel attributes
+  var $panel = this.getPanel($tab)
+  $panel.removeAttribute('role')
+  $panel.removeAttribute('aria-labelledby')
+  $panel.classList.remove(this.jsHiddenClass)
+}
+
+Tabs.prototype.onTabClick = function (e) {
+  e.preventDefault()
+  var $newTab = e.target
+  var $currentTab = this.getCurrentTab()
+  this.hideTab($currentTab)
+  this.showTab($newTab)
+  this.createHistoryEntry($newTab)
+}
+
+Tabs.prototype.createHistoryEntry = function ($tab) {
+  var $panel = this.getPanel($tab)
+  var id = $panel.id
+  $panel.id = ''
```

Comment:
Why do we have to null and restore the panel's `id`? If this is behaviour we definitely need, can we add a comment to explain why?

Reply:
It is so the page doesn't jump when a user clicks a tab (which changes the
hash).